src/Microsoft.Extensions.Hosting.Abstractions/BackgroundService.cs

```diff
@@ -0,0 +1,88 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.Hosting
+{
+    /// <summary>
+    /// Base class for implementing a long running <see cref="IHostedService"/>.
+    /// </summary>
+    public abstract class BackgroundService : IHostedService, IDisposable
+    {
+        private Task _executingTask;
+        private CancellationTokenSource _executionCts;
+        private readonly CancellationTokenSource _startCts = new CancellationTokenSource();
+        private readonly CancellationTokenSource _stopCts = new CancellationTokenSource();
+
+        /// <summary>
+        /// This method is called when the <see cref="IHostedService"/> starts. The implementation should return a task that represents
+        /// the lifetime of the long running operation(s) being performed.
+        /// </summary>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> that fires when the <see cref="IHostedService"/> is stopped.</param>
+        /// <returns>A <see cref="Task"/> that represents the long running operations.</returns>
+        protected abstract Task ExecuteAsync(CancellationToken cancellationToken);
```

Comment:
This CT is non-traditional because it indicates a request for graceful shutdown rather than abortive. Also move to an appropriately named property?

Reply:
I'd prefer if it was part of the method because it's the main use case, "tell me when I'm being stopped". Everything else is edge casey. As for naming, we use tokens to represent all sorts of things now even things that aren't abortive.